storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/GenericResourceAwareSchedulingPriorityStrategy.java

```diff
@@ -0,0 +1,178 @@
+                String resource = entry.getKey();
+                Double requestedAmount = entry.getValue();
+                assignedGenericResources.put(resource, assignedGenericResources.getOrDefault(resource, 0.0) + requestedAmount);
+            }
+            return td;
+        }
+
+        /**
+         * Get a score for the simulated user.  This is used to sort the users, by their highest priority topology.
+         * Only give user guarantees that will not exceed cluster capacity.
+         * Score of each resource type is calculated as: (Requested + Assigned - Guaranteed)/clusterAvailable
+         * The final score is a max over all resource types.
+         * Topology score will fall into the following intervals if:
+         *      User is under quota (guarantee):                    [(-guarantee)/available : 0]
+         *      User is over quota:                                 (0, infinity)
+         * Unfortunately, score below 0 does not guarantee that the topology will be scheduled due to resources fragmentation.
+         * @param availableCpu available CPU on the cluster.
+         * @param availableMemory available memory on the cluster.
+         * @param availableGenericResources available generic resources (other that cpu and memory) in cluster
+         * @param td the topology we are looking at.
+         * @return the score.
+         */
+        protected double getScore(double availableCpu, double availableMemory,
+                                  Map<String, Double> availableGenericResources, TopologyDetails td) {
+            // calculate scores for cpu and memory first
+            double ret = super.getScore(availableCpu, availableMemory, td);
+            if (ret == Double.MAX_VALUE) {
+                return ret;
+            }
+            Map<String, Double> tdTotalRequestedGeneric = td.getTotalRequestedGenericResources();
+            if (tdTotalRequestedGeneric == null) {
+                tdTotalRequestedGeneric = Collections.emptyMap();
+            }
+            for (Map.Entry<String, Double> entry : availableGenericResources.entrySet()) {
+                String resource = entry.getKey();
+                Double available = entry.getValue();
+                if (available <= 0) {
+                    return Double.MAX_VALUE;
+                }
+                Double wouldBeResource = assignedGenericResources.getOrDefault(resource, 0.0)
+                    + tdTotalRequestedGeneric.getOrDefault(resource, 0.0);
+                double thisScore = (wouldBeResource -  guaranteedGenericResources.getOrDefault(resource, 0.0)) / available;
+                ret = Math.max(ret, thisScore);
+            }
+
+            return ret;
+        }
+
+        protected double getScore(double availableCpu, double availableMemory, Map<String, Double> availableGenericResources) {
+            TopologyDetails td = getNextHighest();
```

Comment:
Nit: could we inline this to avoid unnecessary object creation

Reply:
Fixed the inline.

As for `tmpEvictedTopologiesMap`, my concern is that in the the `scheduleTopology()` method, we have a lot of return/break points since we have a pretty complex nested if else structure. Passing the `tmpEvictedTopologiesMap` can make sure the snapshot will be passed to the `evictedTopologiesMap` after this round scheduling finished for whatever reason. 

So we create a new `tmpEvictedTopologiesMap` object before each round of `scheduleTopology()` and passed it over afterward.

Reply:
Thank you - I figured that was why - just wanted to be sure I wasn't missing anything